crates/zed/src/languages/purescript.rs

```diff
@@ -0,0 +1,197 @@
+                    },
+                    {
+                        "fileMatch": [schema_file_match(&paths::KEYMAP)],
+                        "schema": KeymapFile::generate_json_schema(&action_names),
+                    }
+                ]
+            }
+        })
+    }
+}
+
+#[async_trait]
+impl LspAdapter for PurescriptLspAdapter {
+    fn name(&self) -> LanguageServerName {
+        LanguageServerName("purescript-language-server".into())
+    }
+
+    fn short_name(&self) -> &'static str {
+        "purescript"
+    }
+
+    async fn fetch_latest_server_version(
+        &self,
+        _: &dyn LspAdapterDelegate,
+    ) -> Result<Box<dyn 'static + Send + Any>> {
+        Ok(Box::new(
+            self.node
+                .npm_package_latest_version("purescript-language-server")
+                .await?,
+        ) as Box<_>)
+    }
+
+    async fn fetch_server_binary(
+        &self,
+        version: Box<dyn 'static + Send + Any>,
+        container_dir: PathBuf,
+        _: &dyn LspAdapterDelegate,
+    ) -> Result<LanguageServerBinary> {
+        let version = version.downcast::<String>().unwrap();
+        let server_path = container_dir.join(SERVER_PATH);
+
+        if fs::metadata(&server_path).await.is_err() {
+            self.node
+                .npm_install_packages(
+                    &container_dir,
+                    &[("purescript-language-server", version.as_str())],
+                )
+                .await?;
+        }
+
+        Ok(LanguageServerBinary {
+            path: self.node.binary_path().await?,
+            arguments: server_binary_arguments(&server_path),
+        })
+    }
+
+    async fn cached_server_binary(
+        &self,
+        container_dir: PathBuf,
+        _: &dyn LspAdapterDelegate,
+    ) -> Option<LanguageServerBinary> {
+        get_cached_server_binary(container_dir, &*self.node).await
+    }
+
+    async fn installation_test_binary(
+        &self,
+        container_dir: PathBuf,
+    ) -> Option<LanguageServerBinary> {
+        get_cached_server_binary(container_dir, &*self.node).await
+    }
+
+    fn initialization_options(&self) -> Option<serde_json::Value> {
+        Some(json!({
+            "purescript": {
+                "addSpagoSources": true
+            }
+        }))
+    }
+
+    fn workspace_configuration(&self, _workspace_root: &Path, cx: &mut AppContext) -> Value {
+        self.workspace_config
+            .get_or_init(|| Self::get_workspace_config(self.languages.language_names(), cx))
+            .clone()
+    }
+
+    fn language_ids(&self) -> HashMap<String, String> {
+        [("purescript".into(), "purs".into())].into_iter().collect()
+    }
+
+    fn process_diagnostics(&self, l: &mut lsp::PublishDiagnosticsParams) {
+        println!("{:?}", l)
+    }
+}
+
+async fn get_cached_server_binary(
+    container_dir: PathBuf,
+    node: &dyn NodeRuntime,
+) -> Option<LanguageServerBinary> {
+    (|| async move {
```

Comment:
We have an `async_maybe!` macro in our `util` crate that can make this a bit nicer than having the IIFE.

Here's an example of it used in another LSP adapter: https://github.com/zed-industries/zed/blob/5b9cc2619491c33b133254770ec2494010c2d016/crates/zed/src/languages/gleam.rs#L104